### ISSUE TITLE
exporting a type outside of its declaration incorrectly outputs an export

### ISSUE THREAD

**Issue by kiliancs** [alangpierce/sucrase#398]
Using sucrase 3.9.3, this valid TypeScript

```ts
interface Test {

}

type Test2 = number;

export { Test, Test2 };

export default Test;
```

outputs js that exports objects `Test` and `Test2`, which don't exist.

https://sucrase.io/#code=interface%20Test%20%7B%0A%0A%7D%0A%0Atype%20Test2%20%3D%20number%3B%0A%0Aexport%20%7B%20Test%2C%20Test2%20%7D%0A%0Aexport%20default%20Test%3B

**Comment by alangpierce** [comment]
Funny, Babel fails on this one as well, it seems. We elide type-only imports, so probably we just need to do the same for exports.

**Comment by kiliancs** [comment]
I've got a local branch that fixes this. Tests pass except for extra line break, which I'll figure out. Hopefully it's not too terrible. Will make a PR later today or in the morning.

**Comment by kiliancs** [comment]
So I have https://github.com/kiliancs/sucrase/pull/new/unexport_types

This is a very raw commit with what I've been able to figure out how to do. I keep a list of type identifiers, but the function that does this is misplaced (lives with `getNonTypeIdentifiers`) and is actually wrong (there are `export default SomeClass` cases that are incorrectly elided). Also, ideally it wouldn't exist, and we'd just use `getNonTypeIdentifiers`, but it would have to be changed because in the following case, `T` was added to the list of non type identifiers because of the token in the last line:

```ts
type Test = number;
export { Test };
```

I'm sharing this here because I'm not getting much further by myself and in case it's useful, but I suspect it would be faster for everyone if you did it. I've learned a lot about sucrase while doing this, though.

**Comment by kiliancs** [comment]
Ok, I undertand what's going on now. The problem with `export default SomeClass` is due to the fact that `SomeClass` is used as a type elsewhere in the file, and since I'm using a list of type identifiers instead of a list of non identifiers like the one created by `getNonTypeIdentifiers`, the class is not exported.

So I guess this means I'd have to go back to using getNonTypeIdentifiers and adjust it so that an `export { x, y, z }` situation doesn't qualify `x`, `y`, or `z` as non-type identifiers.

**Comment by alangpierce** [comment]
> I suspect it would be faster for everyone if you did it

Makes sense! Yeah, open source is tricky 😄. Currently I'm pretty motivated to maintain this project and can find some time for it, but ideally it would be a little more of a collaboration so that it could continue to stay up-to-date if I get burnt out or move to something else. But it does seem like for a lot of the more complex tasks, it's easier if I just do it, so I'd be happy to take this one on. Thanks for being involved with the project! 😄

**Comment by kiliancs** [comment]
I'll keep trying and I'll be looking at your solutions, so hopefully soon I can contribute. Sucrase has incredible potential for large typescript apps, but I think it still has a long way to go in terms of TypeScript compatibility. Maybe it would be useful to actively test it against even more large codebases. I think we'll still find more stuff as we fix the ones I'm finding by running it in ours.

**Comment by alangpierce** [comment]
👍 thanks for all of the bug reports! TBH it felt like it was really stable, but I guess I was overconfident. My codebase at work is about 500,000 lines of code, but only about 50,000 of that is TypeScript, so makes sense that it would miss some cases. If you know big open source TypeScript codebases to use as a test case, would be good to hear! Unfortunately Sucrase can't compile `namespace`s (also a limitation of Babel and TS with `isolatedModules`), so the TypeScript compiler itself can't be tested. Ideally I'd be testing real application code, but there aren't a lot of big applications on GitHub.

**Comment by alangpierce** [comment]
Ooh, this one is interesting, I guess exporting a type like that is disallowed by the `isolatedModules` flag.

<img width="587" alt="screen shot 2019-01-13 at 8 20 22 pm" src="https://user-images.githubusercontent.com/211605/51096360-cdeabd80-1770-11e9-8c29-ae423052dc18.png">

This is also why Babel doesn't support it: https://github.com/babel/babel/issues/6065

As mentioned in the README, Sucrase is only intended to work with `isolatedModules`-compatible code, since it fundamentally doesn't support cross-file compilation. That said, the TypeScript compiler manages to elide the exports here, and it certainly seems doable to get this working in Sucrase. I think`getNonTypeIdentifiers` actually won't work because really we want to look at *declarations* and `getNonTypeIdentifiers` only focuses on *usages*. For example, `const x = 0; export {x};` vs `type T = number; export {T};` are only different in how the identifier is declared.

**Comment by kiliancs** [comment]
That TS1205 error seems incorrect since that's not a re-export. Related issue https://github.com/Microsoft/TypeScript/issues/28481. Though one could argue exporting types with `isolatedModules` makes no sense, why is `export type Test2 = number` allowed? I suspect even with `isolatedModules` there are valid cases for exporting types, just not re-exporting.

Ultimately, Sucrase has to deal with the limitations derived of not knowing if something imported from another file is a type, but I don't think this case here, like you say, falls within that limitation.

Even type re-exports, at a practical level, could be consider OK if the goal is a fast dev build. If I import a type and then export it, it will compile some code exporting with value undefined. If it's a type, TypeScript, which I'd use only for type analysis, would prevent me from using that type as a runtime value anywhere else.

I think it should be possible and good for most users to support a setup with namespaces, split enums, etc provided (which can be done through tslint, for example, but it might be ok to just let Sucrase complain), and with TypeScript doing the type analysis, with `isolatedModules` disabled.

I understand right now Sucrase only commits to support whatever `isolatedModules` allows, but at the same time it seems it's so close to being able to fully support any modern typescript codebase, provided no namespaces nor split enums, which are discouraged anyway. Maybe there are more things that I'm missing, but right now it seems this issue is the only thing preventing this sort of holy grail, super fast, super safe, setup for dev.

**Comment by alangpierce** [comment]
I just released a fix for this in 3.10.0.
